toffy/fov_watcher.py

```diff
@@ -0,0 +1,217 @@
+
+        return False, fov_name
+
+    def check_completion(self) -> dict:
+        """Condenses internal dictionary to show which fovs have finished
+
+        Returns:
+            dict
+        """
+        return {k: all(self.completion[k].values) for k in self.completion}
+
+
+class FOV_EventHandler(FileSystemEventHandler):
+    """File event handler for FOV files
+
+    Attributes:
+        run_folder (str):
+            path to run folder
+        watcher_out (str):
+            folder to save all callback results + log file
+        run_structure (RunStructure):
+            expected run file structure + completion status
+        per_fov (list):
+            callbacks to run on each fov
+        per_run (list):
+            callbacks to run over the entire run
+        panel (tuple | pd.DataFrame):
+            masses to extract
+    """
+    def __init__(self, run_folder: str, per_fov: List[Callable[[xr.DataArray, str], None]],
+                 per_run: List[Callable[[str, str], None]],
+                 panel: Union[Tuple, pd.DataFrame] = (0.3, 0.0)):
+        """Initializes FOV_EventHandler
+
+        Args:
+            run_folder (str):
+                path to run folder
+            per_fov (list):
+                callbacks to run on each fov
+            per_run (list):
+                callbacks to run over the entire run
+            panel (tuple | pd.DataFrame):
+                masses to extract
+        """
+        super().__init__()
+        self.run_folder = run_folder
+
+        self.watcher_out = os.path.join(run_folder, 'watcher_outs')
+
+        if not os.path.exists(self.watcher_out):
+            os.makedirs(self.watcher_out)
+
+        # create run structure
+        self.run_structure = RunStructure(run_folder)
+
+        self.per_fov = per_fov
+        self.per_run = per_run
+        self.panel = panel
+
+        for root, dirs, files in os.walk(run_folder):
+            for name in files:
+                self.on_created(FileCreatedEvent(os.path.join(root, name)))
+
+    def on_created(self, event: FileCreatedEvent):
+        """Handles file creation events
+
+        If FOV structure is completed, all callbacks in `per_fov` will be run over the data.
+        This function is automatically called; users generally shouldn't call this function
+
+        Args:
+            event (FileCreatedEvent):
+                file creation event
+        """
+        super().on_created(event)
+
+        # check if what's created is in the run structure
+        fov_ready, point_name = self.run_structure.check_run_condition(event.src_path)
+        if fov_ready:
+            log_file_path = os.path.join(self.watcher_out, 'log.txt')
+            logf = open(log_file_path, 'a')
+
+            logf.write(
+                f'{datetime.now().strftime("%d/%m/%Y %H:%M:%S")} -- '
+                f'Extracting {point_name}'
+            )
+
+            # extract data
+            img_data = \
```

Comment:
If we have multiple functions that require extracted data then this makes sense, otherwise it may be easier to have each watcher function handle its own processing

Reply:
honestly, even just for the 'bonus' of having each watcher function have its own extraction parameters would probably be handy (intensities y/n, integration range checks, etc).  I'll change it to pass the run folder + fov name instead.